classes/Cart.php

```diff
@@ -744,43 +744,45 @@ public function getProducts($refresh = false, $id_product = false, $id_country =
         Product::cacheProductsFeatures($products_ids);
         Cart::cacheSomeAttributesLists($pa_ids, $this->id_lang);
 
-        $this->_products = array();
```

Comment:
You can keep this line and remove 748 & 785 to prevent duplicate.

Reply:
No you can't.

Because on the line 758 you call $this->getCartRules(CartRule::FILTER_ACTION_GIFT) but in nested action calls you will call Cart getProducts method again, but if you have defined $this-_products = array() before then you will always have empty product lists.

Reply:
Thanks for details, I ping our Cart master to let him check if everything is ok!